### ISSUE TITLE
Nightly failures with desul atomics

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#4136]
Some nightly tests began failing following merge of PR #3247 

* intel/17 (compilation error):
```
20:39:35 /home/jenkins/node/workspace/kokkos_osrhel7_intel_1701_openmp_pthread_serial/kokkos/core/src/desul/atomics/Generic.hpp(62): error: type name is not allowed
20:39:35   constexpr DESUL_FUNCTION typename std::enable_if<may_exit_early<Op, Scalar1, Scalar2>{}, bool>::type
```

Sample reproducer (blake testbed)
```
module load sems-env sems-cmake/3.17.1 sems-intel/17.0.1
module swap gcc/4.9.3 gcc/6.4.0

$KOKKOS_PATH/generate_makefile.bash --with-devices=Serial  --compiler=icpc --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14"  --kokkos-path=$KOKKOS_PATH --no-examples --with-options=enable_large_mem_tests
```

* clang/7+cuda/9.2 (-Werror):
```
18:23:18 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkos-dev_clang_701_cuda_92/kokkos/core/src/desul/atomics/Compare_Exchange_CUDA.hpp:138:22: error: unused parameter 'compare' [-Werror,-Wunused-parameter]
18:23:18     T* const dest, T compare, T value, MemoryOrderAcquire, MemoryScope) {
```

Sample reproducer (kokkos-dev)
```
module load sems-env sems-cmake/3.17.1 sems-clang/7.0.1 sems-cuda/9.2

$KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,OpenMP --arch=Kepler35 --compiler=clang++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --with-cuda=/projects/sems/install/rhel7-x86_64/sems/compiler/cuda/9.2/base --kokkos-path=$KOKKOS_PATH --with-options=disable_deprecated_code
```

* armclang/20.1 (openmp.task_fib_deprecated test)
```
18:51:18 1: [ RUN      ] openmp.task_fib_deprecated
18:51:18 1: TaskScheduler task has negative reference count or is incomplete
18:51:18  1/41 Test  #1: KokkosCore_UnitTest_OpenMP ...................Child aborted***Exception:  31.52 sec
```
Edit: This openmp unit test also fails on Power9 in Cuda+OpenMP build compiled with cuda/10.1.243 and gcc/7.4.0

Sample reproducer (mayer testbed)
```
module load cmake/3.17.1 arm/20.1

$KOKKOS_PATH/generate_makefile.bash --with-devices=OpenMP --arch=ARMV8_THUNDERX2 --compiler=armclang++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --ldflags=""  --kokkos-path=$KOKKOS_PATH --no-examples --with-options=enable_large_mem_tests
```

* KokkosKernels and Trilinos (likely all builds)
```
21:52:55 /data/jenkins-new/workspace/KokkosKernels_semsrhel7gpu01_cuda111_gcc830-dbg-boundscheck/kokkos-kernels/src/sparse/KokkosSparse_CrsMatrix.hpp(728): error: namespace "Kokkos" has no member "atomic_assign"
```

I'll open a separate issue for KokkosKernels (there should be other locations errors show up with use of atomics beside the message above). In the short term we can consider disabling `KOKKOS_ENABLE_IMPL_DESUL_ATOMICS` if necessary until KokkosKernels is updated for desul atomics, adding @lucbv so he is in the loop

@crtrott should we consider adding guards to set the default option to disable desul atomics for Trilinos https://github.com/kokkos/kokkos/blob/01766966210f010b1499d8ad796af681e59ed671/cmake/kokkos_enable_options.cmake#L62 , or is the plan for Trilinos to be switched over to use desul atomics before the 3.5 release?

**Comment by crtrott** [comment]
The plan is to switch, so we can get all the failures, but turn it off for now in the Trilinos and KokkosKernels tests for jenkins until the plain Kokkos tests pass. I will try and fix all of it tomorrow.

**Comment by ndellingwood** [comment]
Adding info for xl nightly compilation error:

* xl/16.1.1 serial build (white testbed):
```
01:10:17 CMakeFiles/KokkosCore_UnitTest_Serial1.dir/serial/TestSerial_Atomics.cpp.o: In function `void Kokkos::parallel_for<TestAtomic::CASFunctor<Kokkos::complex<float>, Kokkos::Serial> >(unsigned long, TestAtomic::CASFunctor<Kokkos::complex<float>, Kokkos::Serial> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
01:10:17 /home/jenkins/white-new/workspace/Kokkos_develop_white_xl_1611_serial/TestAll_2021-07-02_01.02.36/ibm/16.1.1/Serial-release/core/unit_test/serial/TestSerial_Atomics.cpp:(.text+0x3e398): undefined reference to `__atomic_compare_exchange'
01:10:17 CMakeFiles/KokkosCore_UnitTest_Serial1.dir/serial/TestSerial_Atomics.cpp.o: In function `void Kokkos::parallel_for<TestAtomic::ExchFunctor<Kokkos::complex<float>, Kokkos::Serial> >(unsigned long, TestAtomic::ExchFunctor<Kokkos::complex<float>, Kokkos::Serial> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
01:10:17 /home/jenkins/white-new/workspace/Kokkos_develop_white_xl_1611_serial/TestAll_2021-07-02_01.02.36/ibm/16.1.1/Serial-release/core/unit_test/serial/TestSerial_Atomics.cpp:(.text+0x3ea98): undefined reference to `__atomic_exchange'
01:10:17 CMakeFiles/KokkosCore_UnitTest_Serial1.dir/serial/TestSerial_Atomics.cpp.o: In function `Kokkos::Impl::ParallelReduce<TestAtomic::CASFunctorReduce<Kokkos::complex<float>, Kokkos::Serial>, Kokkos::RangePolicy<Kokkos::Serial>, Kokkos::InvalidType, Kokkos::Serial>::execute() const':
01:10:17 /home/jenkins/white-new/workspace/Kokkos_develop_white_xl_1611_serial/TestAll_2021-07-02_01.02.36/ibm/16.1.1/Serial-release/core/unit_test/serial/TestSerial_Atomics.cpp:(.text+0x44348): undefined reference to `__atomic_compare_exchange'
01:10:17 CMakeFiles/KokkosCore_UnitTest_Serial1.dir/serial/TestSerial_Atomics.cpp.o: In function `Kokkos::Impl::ParallelReduce<TestAtomic::ExchFunctorReduce<Kokkos::complex<float>, Kokkos::Serial>, Kokkos::RangePolicy<Kokkos::Serial>, Kokkos::InvalidType, Kokkos::Serial>::execute() const':
01:10:17 /home/jenkins/white-new/workspace/Kokkos_develop_white_xl_1611_serial/TestAll_2021-07-02_01.02.36/ibm/16.1.1/Serial-release/core/unit_test/serial/TestSerial_Atomics.cpp:(.text+0x44a98): undefined reference to `__atomic_exchange'
01:10:17 make[2]: *** [core/unit_test/KokkosCore_UnitTest_Serial1] Error 1
```

Sample reproducer
```
module load cmake/3.19.3 ibm/xl/16.1.1 gcc/7.2.0

$KOKKOS_PATH/generate_makefile.bash --with-devices=Serial --arch=Power8,Pascal60 --compiler=xlC --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized -isystem /home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0/powerpc64le-unknown-linux-gnu -qgcc_cpp_stdinc=/home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0 -qxflag=disable__cplusplusOverride -I/home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0/powerpc64le-unknown-linux-gnu" --cxxstandard="14" --ldflags=""  --kokkos-path=$KOKKOS_PATH --no-examples --with-options=enable_large_mem_tests
```

**Comment by ndellingwood** [comment]
We still need fixes for the armclang and xl issues

**Comment by ndellingwood** [comment]
@crtrott After the first wave of fixes went in the Cuda_OpenMP build on Power9 resumed passing (the openmp.task_fib_deprecated test previously failed), but a new OpenMP test failed last even though there were no new merged changes:

```
03:28:44 1: [ RUN      ] openmp.impl_shared_alloc
03:28:44 1: Kokkos::Impl::SharedAllocationRecord failed is_sane: rec(0x200004086ed0){ m_count(0) m_root(0x000012236590) m_next(0x000000000000) m_prev(0x20002c07afe0) m_next->m_prev(0x000000000000) m_prev->m_next(0x200004086ed0) }
03:28:44 1: Kokkos::Impl::SharedAllocationRecord failed is_sane: rec(0x200004086ed0){ m_count(0) m_root(0x000012236590) m_next(0x000000000000) m_prev(0x20002c07afe0) m_next->m_prev(0x000000000000) m_prev->m_next(0x200004086ed0) }
03:28:44 1: Kokkos::Impl::SharedAllocationRecord failed is_sane: rec(0x200004086ed0){ m_count(0) m_root(0x000012236590) m_next(0x000000000000) m_prev(0x20002c07afe0) m_next->m_prev(0x000000000000) m_prev->m_next(0x200004086ed0) }

...
03:28:44 1: Kokkos::Impl::SharedAllocationRecord failed is_sane: rec(0x200004086ed0){ m_count(1953719668) m_root(0x000012236590) m_next(0x200004086e80) m_prev(0x2000340d2300) m_next->m_prev(0x000000000000) m_prev->m_next(0x200004086ed0) }
03:28:44  1/51 Test  #1: KokkosCore_UnitTest_OpenMP ...................***Exception: SegFault 76.36 sec
```

Reproducer (weaver test bed):
```
module load cmake/3.19.3 cuda/10.1.243 ibm/xl/16.1.1 gcc/7.4.0

$KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,OpenMP --arch=Power9,Volta70 --compiler=$KOKKOS_PATH/bin/nvcc_wrapper --cxxflags="-g -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --with-cuda=/home/projects/ppc64le-pwr9-nvidia/cuda/10.1.243 --kokkos-path=$KOKKOS_PATH --with-cuda-options=enable_lambda,uvm,rdc --boundscheck --no-examples --with-options=enable_large_mem_tests --debug
```

**Comment by crtrott** [comment]
The armclang thing should be fixed, I reran Mayer jenkins jobs and they passed (note it was no armclang per se, the same problem happened with gcc, since the root cause is a missing fence which manifests on architectures with less strong memory ops ordering than X86).

**Comment by ndellingwood** [comment]
@crtrott thanks, looks like good progress, a couple more failures remain:

* Undefined reference errors compiling with XL (host backends and as host-compiler with Cuda builds): https://github.com/kokkos/kokkos/issues/4136#issuecomment-873200530

* sems-intel/17 unit test failure in scatterview

```
21:07:52 39: [ RUN      ] openmp.scatterview
21:07:52 39/41 Test #39: KokkosContainers_UnitTest_OpenMP .............***Exception: Illegal  0.57 sec
```

That one crept in after the intel compiler desul fixes went in
```
module load sems-env sems-cmake/3.17.1 sems-intel/17.0.1

$KOKKOS_PATH/generate_makefile.bash --with-devices=OpenMP  --compiler=icpc --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --kokkos-path=$KOKKOS_PATH --no-examples --with-options=enable_large_mem_tests
```

**Comment by crtrott** [comment]
The sems-intel/17 one is actually older: #4113 but the start of that guy was masked by #3936

**Comment by ndellingwood** [comment]
@crtrott just to clarify, the sems-intel/17.0.1 nightly was passing prior to the desul merge (last passed 6/30/2021 https://jenkins-son.sandia.gov/view/Kokkos/job/kokkos_osrhel7_intel_1701_openmp_pthread_serial/19/console). After the fixes for compilation errors went in the openmp.scatterview test began failing https://jenkins-son.sandia.gov/view/Kokkos/job/kokkos_osrhel7_intel_1701_openmp_pthread_serial/22/
The #4113 failure (with intel/17.0.4 on blake) showed a couple weeks ago and masked by #3936 like you mentioned, but I didn't think this is connected to the scatterview failure since that is recent?

**Comment by crtrott** [comment]
Oh ok, though i think its the same issue. Its connected to something with atomics though, but the original thing in the random unit test exists before desul. Let me check the scatter view thing. Note its not fully deterministic ..

**Comment by PhilMiller** [comment]
Fix released in 3.6
